src/main/java/teammates/ui/controller/InstructorCourseInstructorAddAction.java

```diff
@@ -162,73 +197,158 @@ private void updateInstructorWithSectionLevelPrivileges(String courseId, Instruc
+            HashMap<String, Boolean> isSectionSpecialMappings) {
+        HashMap<String, List<String>> specialSectionsInSectionGroups = new HashMap<String, List<String>>();
         if (instructorToAdd.role.equals(Const.InstructorPermissionRoleNames.INSTRUCTOR_PERMISSION_ROLE_CUSTOM)) {
-            for (int i = 0; i < sectionNames.size(); i++) {
-                String setSectionGroupStr = getRequestParamValue("is" + Const.ParamsNames.INSTRUCTOR_SECTION_GROUP + i + "set");
-                boolean isSectionGroupSpecial = "true".equals(setSectionGroupStr);
-                for (int j = 0; j < sectionNames.size(); j++) {
-                    String valueForSectionName =
-                            getRequestParamValue(Const.ParamsNames.INSTRUCTOR_SECTION_GROUP + i
-                                                 + Const.ParamsNames.INSTRUCTOR_SECTION + j);
-                    if (isSectionGroupSpecial && valueForSectionName != null && sectionNamesTable.containsKey(valueForSectionName)) {
-                        if (sectionNamesMap.get(Const.ParamsNames.INSTRUCTOR_SECTION_GROUP + i) == null) {
-                            sectionNamesMap.put(Const.ParamsNames.INSTRUCTOR_SECTION_GROUP + i, new ArrayList<String>());
-                        }
-                        sectionNamesMap.get(Const.ParamsNames.INSTRUCTOR_SECTION_GROUP + i).add(valueForSectionName);
-                        sectionNamesTable.put(valueForSectionName, true);
-                    }
+            getSectionsWithSpecialPrivilegesForCustomInstructor(sectionNames, isSectionSpecialMappings,
+                                                                specialSectionsInSectionGroups);
+        }
+        return specialSectionsInSectionGroups;
+    }
+    
+    /**
+     * Gets the sections that are special for the custom instructor to be added.
+     * Prereq: the added instructor must be given a custom role.
+     * 
+     * @param sectionNames                   List of section names in the course.
+     * @param isSectionSpecialMappings       Mapping of names of sections to boolean values indicating if they are special.
+     *                                           This will be modified within the method.
+     * @param specialSectionsInSectionGroups Mapping of section group names to the special sections that they contain.
+     *                                           This will be modified within the method.
+     */
+    private void getSectionsWithSpecialPrivilegesForCustomInstructor(List<String> sectionNames,
+            HashMap<String, Boolean> isSectionSpecialMappings,
+            HashMap<String, List<String>> specialSectionsInSectionGroups) {
+        for (int i = 0; i < sectionNames.size(); i++) {
+            String sectionGroupIsSetStr = getRequestParamValue("is"
+                                                  + Const.ParamsNames.INSTRUCTOR_SECTION_GROUP + i + "set");
+            boolean isSectionGroupSpecial = "true".equals(sectionGroupIsSetStr);
```

Comment:
It was done this before, but `Boolean.parseBoolean(sectionGroupIsSetStr)` should be better.